source/ColumnChooserPanel.cpp

```diff
@@ -0,0 +1,129 @@
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ColumnChooserPanel.h"
+
+#include "FillShader.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "GameData.h"
+#include "Information.h"
+#include "Interface.h"
+#include "Preferences.h"
+#include "Sprite.h"
+#include "SpriteSet.h"
+#include "SpriteShader.h"
+#include "UI.h"
+#include "text/WrappedText.h"
+
+#include <numeric>
+
+using namespace std;
+
+ColumnChooserPanel::ColumnChooserPanel(const vector<PlayerInfoPanel::SortableColumn> &columns,
+	InfoPanelState *panelState)
+	: columns(columns), panelState(panelState)
+{
+	SetInterruptible(false);
+}
+
+
+
+void ColumnChooserPanel::Draw()
+{
+	if(GetUI()->IsTop(this))
+		DrawBackdrop();
+
+	zones.clear();
+
+	Information info;
+	const Interface *columnChooser = GameData::Interfaces().Get("columns menu");
+	info.SetCondition("columns menu open");
+	columnChooser->Draw(info, this);
+
+	const Font &font = FontSet::Get(14);
+	const Color * const dim = GameData::Colors().Get("dim");
+	const Color * const medium = GameData::Colors().Get("medium");
+	const Color * const bright = GameData::Colors().Get("bright");
+	const Sprite *box[] = {SpriteSet::Get("ui/unchecked"), SpriteSet::Get("ui/checked")};
+
+	Point topLeft(270., -280.);
+	const Point boxSize = Point(box[0]->Width(), box[0]->Height());
+	static const Point ROW_ADVANCE(0., 20.);
+	const set<const string> visibleColumns = panelState->VisibleColumns();
+	auto isVisible = [&](string name){ return visibleColumns.find(name) != visibleColumns.end(); };
+	const float availableWidth = 727 - accumulate(columns.begin(), columns.end(), 0,
+		[&](int acc, PlayerInfoPanel::SortableColumn column) {
+			return acc + (isVisible(column.name) ? column.layout.width : 0);
+		});
+	for(PlayerInfoPanel::SortableColumn &column : columns)
+	{
+		Rectangle zoneBounds = Rectangle::FromCorner(topLeft, Point(220., ROW_ADVANCE.Y()));
+
+		bool visible = isVisible(column.name);
+		bool enabled = visible || column.layout.width <= availableWidth;
+		bool hover = zoneBounds.Contains(hoverPoint);
+
+		Rectangle spriteBounds = Rectangle::FromCorner(topLeft, boxSize);
+		SpriteShader::Draw(box[visible], spriteBounds.Center());
+
+		Point textPos = topLeft + Point(box[0]->Width(), 2.);
+		font.Draw(column.checkboxLabel, textPos, enabled ? hover ? *bright : *medium : *dim);
+
+		if(enabled)
+			zones.emplace_back(ClickZone(zoneBounds, column.name));
+
+		topLeft = topLeft + ROW_ADVANCE;
+	}
+}
+
+
+
+bool ColumnChooserPanel::AllowsFastForward() const noexcept
+{
+	return true;
+}
+
+
+
+bool ColumnChooserPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command, bool isNewPress)
+{
+	if(key == 'a' || key == 'n')
+		GetUI()->Pop(this);
+	else
+		return false;
+
+	return true;
+}
+
+
+
+bool ColumnChooserPanel::Click(int x, int y, int /* clicks */)
+{
+	Point mouse(x, y);
+	for(auto &zone : zones)
+		if(zone.Contains(mouse))
+			return panelState->ToggleColumn(zone.Value()), true;
```

Comment:
What's happening here?

Reply:
Every frame, the valid click zone rectangles are re-computed, re-added to the `zones` list, an then iterated though here to check if the mouse click is within one of them (zones are assumed not to overlap—if they do, only the first one is found). If it is, the corresponding checkbox state is 'toggled' and Click() returns true to prevent further processing. If no zone contains the mouse location, the loop ends and the PC falls through to the end of the function where false is returned, and the calling function moves up the panel heirarchy to see if the panel below this one can hande a mouse click at that location.

Reply:
no, I mean what does the `, true` do?

Reply:
That's the return value for the function.

Reply:
then what does the `return panelState->ToggleColumn(zone.Value())` do?

Reply:
Toggles the state of the checkbox.
Do you understand the comma operator in C?

(Also it's a penalty shootout so i won't be paying much attention for 19 mins)

Reply:
```suggestion
		{
			panelState->ToggleColumn(zone.Value());
			return true;
		}
```

Reply:
(Yay, we won!)

I have not worked extensively on a C++ project since around 2005. I am learning a lot about C++11, 14, 17 and 20 from this project.

This suggestion feels like you're suggesting I replace `x < y` with `y >= x` because you're not familiar with the less-than sign.

I chose this representation because it feels and looks much cleaner to me than the floating indented brackets that would arise from the form you're suggesting.

May I suggest instead that you take this as a learning opportunity. The comma is core and useful feature of C/C++, with which you should be familiar.

Reply:
I'm not saying we should avoid it because people may not be familiar with it. But it's basically the same thing as putting two semicolon-separated statements on a single line, which we don't allow, because it's harder to read.

Reply:
Not quite, as that would end the if-block prematurely. There are also some existing places where we have multiple statements on a single line seperated by semi-colons, such as for-statements.
If you really object, I will change it, but I do think it looks a lot nicer the way it currently is.

Reply:
The style guide says this:

> Do not put multiple expression on the same line.
Except, of course, for the three individual clauses contained in a for loop.